service/src/MSRIOGroup.cpp

```diff
@@ -395,11 +399,114 @@ namespace geopm
                                                    IOGroup::M_UNITS_WATTS,
                                                    agg_function(msr_name),
                                                    ps.description + "\n    alias_for: " + ps.msr_name + " rate of change",
-                                                   IOGroup::M_SIGNAL_BEHAVIOR_VARIABLE};
+                                                   IOGroup::M_SIGNAL_BEHAVIOR_VARIABLE,
+                                                   string_format_double};
             }
         }
     }
 
+    void MSRIOGroup::register_pcnt_scalability_signals(void)
+    {
+        // Tracking for intermediate signals that will not
+        // be exposed to the user but will be used to 
+        // generate user visible signals
+        std::map<std::string, signal_info> signal_hidden;
+
+        // register time signal; domain board
+        std::string time_name = "MSR::TIME";
+        std::shared_ptr<Signal> time_sig = std::make_shared<TimeSignal>(m_time_zero, m_time_batch);
+        signal_hidden[time_name] = {std::vector<std::shared_ptr<Signal> >({time_sig}),
+                                    GEOPM_DOMAIN_CPU,
+                                    IOGroup::M_UNITS_SECONDS,
+                                    Agg::select_first,
+                                    "Time in seconds used to calculate pcnt_rate, acnt_rate",
+                                    IOGroup::M_SIGNAL_BEHAVIOR_VARIABLE,
+                                    string_format_double};
+        int derivative_window = 8;
+        double sleep_time = 0.005;  // 5000 us
+
+        // Mapping of high-level signal name to description and
+        // underlying CNT MSRs.  The domain will match that of the
+        // CNT signals.
+        struct cnt_data
+        {
+            std::string cnt_name;
+            std::string description;
+            std::string msr_name;
+        };
+        std::vector<cnt_data> cnt_signals {
+            {"MSR::PCNT_RATE",
+                    "Average cpu pcnt rate over 8 control loop iterations (40ms if using geopmread)",
+                    "MSR::PPERF:PCNT"},
+            {"MSR::ACNT_RATE",
+                    "Average cpu acnt rate over 8 control loop iterations (40ms if using geopmread)",
+                    "MSR::APERF:ACNT"}
+        };
+        for (const auto &ps : cnt_signals) {
```

Comment:
Some comments about this block would help: specifically that you're taking the derivative of PCNT and ACNT rate deltas would help explain what's going on.

Then later you're taking the ratio (DivisionSignal) of the 2 rate deltas.